src/main/java/com/fakng/fakngagrgtr/parser/google/GoogleParser.java

```diff
@@ -9,57 +15,88 @@
-        super(webClient);
+    public GoogleParser(WebClient webClient, LocationCache locationCache, CompanyRepository companyRepository,
+                        LocationRepository locationRepository, @Value("${url.google}") String url) {
+        super(webClient, locationCache, companyRepository, locationRepository);
         this.url = url;
-        this.google = companyRepository.findByTitle(GOOGLE_NAME).orElse(null);
+    }
+
+    @PostConstruct
+    public void init() {
+        google = companyRepository.findByTitle(GOOGLE_NAME).orElseThrow(() -> new IllegalStateException(String.format("There is no %s company present in DB", GOOGLE_NAME)));
+        google.getLocations().forEach(location -> locationCache.putIfAbsent(location.getCity(), location.getCountry(), location));
     }
 
     @Override
     protected List<Vacancy> getAllVacancies() throws Exception {
-        ResponseDTO firstPage = getPage(1);
+        ResponseDto firstPage = getPage(1);
         int lastPage = firstPage.getCount() / firstPage.getPageSize() + 1;
-        List<Vacancy> vacancies = new ArrayList<>();
-        for (int index = 2; index <= lastPage; index += 1) {
-            List<VacancyDTO> jobs = getPage(index).getJobs();
-            vacancies.addAll(jobs.stream().map(this::createVacancy).toList());
+        List<Vacancy> allVacancies = new ArrayList<>(processPageResponse(firstPage));
+        for (int idx = 2; idx <= lastPage; idx++) {
+            allVacancies.addAll(processPageResponse(getPage(idx)));
         }
-        return vacancies;
+        return allVacancies;
+    }
+
+    private List<Vacancy> processPageResponse(ResponseDto response) {
+        return response.getJobs().stream()
+                .map(this::createVacancy)
+                .toList();
     }
 
-    private Vacancy createVacancy(VacancyDTO dto) {
+    private Vacancy createVacancy(VacancyDto dto) {
         Vacancy vacancy = new Vacancy();
         vacancy.setId(parseVacancyId(dto.getId()));
         vacancy.setTitle(dto.getTitle());
         vacancy.setUrl(dto.getApplyUrl());
         vacancy.setCompany(google);
-        if (dto.getPublishDate() != null) {
-            vacancy.setAddDate(parseLocalDateTime(dto.getPublishDate()));
-        }
-        vacancy.setLocation(null); // dto.locations?
+        processLocations(vacancy, dto.getLocations());
         vacancy.setDescription(generateFullDescription(dto));
         return vacancy;
     }
 
+    private void processLocations(Vacancy vacancy, List<LocationDto> locations) {
+        locations.forEach(location -> {
+            String locationKey = locationCache.getLocationKey(location.getCity(), location.getCountryCode());
+            if (locationCache.contains(locationKey)) {
+                vacancy.addLocation(locationCache.get(locationKey));
+            } else {
+                saveInDbAndCache(vacancy, locationKey, location);
```

Comment:
Check the location contains in cache on method start and save to database if need. Then set it to vacancy. Remove vacancy.addLocation from saveInDbAndCache method.